src/expr/impl/src/scalar/case.rs

```diff
@@ -106,6 +107,132 @@ impl Expression for CaseExpression {
+            operand,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Expression for ConstantLookupExpression {
+    fn return_type(&self) -> DataType {
+        self.return_type.clone()
+    }
+
+    async fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
+        let input_len = input.capacity();
+        let mut builder = self.return_type().create_array_builder(input_len);
+
+        // Evaluate the input DataChunk at first
+        let eval_result = self.operand.eval(input).await?;
+
+        for i in 0..input_len {
+            let datum = eval_result.datum_at(i);
+            let (row, vis) = input.row_at(i);
+
+            // Check for visibility
+            if !vis {
+                builder.append_null();
+                continue;
+            }
+
+            // Note that the `owned_row` here is extracted from input
+            // rather than from `eval_result`
+            let owned_row = row.into_owned_row();
+
+            if let Some(expr) = self.arms.get(datum.as_ref().unwrap()) {
+                builder.append(expr.eval_row(&owned_row).await.unwrap().as_ref());
+            } else {
+                // Otherwise this should goes to the fallback arm
+                // The fallback arm should also be const
+                if let Some(ref fallback) = self.fallback {
+                    builder.append(fallback.eval_row(&owned_row).await.unwrap().as_ref());
+                } else {
+                    builder.append_null();
+                }
+            }
+        }
+
+        Ok(Arc::new(builder.finish()))
+    }
+
+    async fn eval_row(&self, input: &OwnedRow) -> Result<Datum> {
+        let datum = self.operand.eval_row(input).await?;
+
+        if let Some(expr) = self.arms.get(datum.as_ref().unwrap()) {
+            expr.eval_row(input).await
+        } else {
+            let Some(ref expr) = self.fallback else {
+                return Ok(None);
+            };
+            expr.eval_row(input).await
+        }
+    }
+}
+
+#[build_function("constant_lookup(...) -> any", type_infer = "panic")]
+fn build_constant_lookup_expr(
+    return_type: DataType,
+    children: Vec<BoxedExpression>,
+) -> Result<BoxedExpression> {
+    if children.is_empty() {
+        bail!("children expression must not be empty for constant lookup expression");
+    }
+
+    let mut children = children;
+
+    let operand = children.remove(0);
+
+    let mut arms = HashMap::new();
+
+    // Build the `arms` with iterating over `when` & `then` clauses
+    let mut iter = children.into_iter().array_chunks();
+    for [when, then] in iter.by_ref() {
+        let Ok(Some(s)) = when.eval_const() else {
+            bail!("expect when expression to be const");
+        };
+        arms.insert(s, then);
+    }
+
+    let fallback = if let Some(else_clause) = iter.into_remainder().unwrap().next() {
+        if else_clause.return_type() != return_type {
+            bail!("Type mismatched between else and case.");
+        }
```

Comment:
Is this enforced by the binder? Though doing the check again is not bad.

Reply:
I don't think this is enforced by the binder (i.e., during `bind_case`), also not in `infer_type_for_special`. And we have the same check for normal case-when pass, so maybe we could keep this check here? 🤔️